server/src/server.js

```diff
@@ -645,6 +645,14 @@ app.get("/images/:imageid", function (req, res) {
         anonAnalytics.pageview(analyticsUrl).send();
       }
       res.header("Content-Type", obj.contentType);
+      if (req.query.filename) {
+        let extension = "." + obj.contentType.split("/").pop();
+        let filename= req.query.filename;
+        if (! filename.endsWith(extension)) {
+          filename = filename + extension;
+        }
+        res.header("Content-Disposition", `attachment; filename="${filename}"`)
+      }
```

Comment:
You wouldn't have to pass anything around in the model. The server would take the filename and the salt and generate the hash, and compare it with the hash provided by the client.

If the button styles apply to an anchor tag and we switch to using the download attribute, we'd still have to include both the uuid and the filename in the url (because the filename isn't enough to uniquely identify the image, and the image uuid isn't enough information to generate the filename), so the same attack would still work.